### PR TITLE
add string overloads to outbuffer.d

### PR DESCRIPTION
**Pull request by WalterBright**
strlen() must die

### CODE REVIEW

**Comment by yebblies** [comment]
Why not overload?

**Comment by WalterBright** [comment]
string literals match both * and []

**Comment by yebblies** [comment]
They will prefer a match with `immutable(char)[]`

**Comment by WalterBright** [comment]
There are a lot of const(char)[] in dmd.

**Comment by yebblies** [comment]
So? Three overloads should cover all cases:
```
void writestring(const(char)*);
void writestring(const(char)[]);
void writestring(string);
```
With no changes to the caller code for string literals.

**Comment by WalterBright** [comment]
You are right. Will fix. This is much better.

**Comment by WalterBright** [comment]
@yebblies ping

File: src/root/outbuffer.d [hunk 1]
```diff
@@ -91,6 +91,16 @@ struct OutBuffer
         write(string, strlen(string));
     }
 
+    void writestring(const(char)[] s) nothrow
+    {
+        write(s.ptr, s.length);
+    }
+
+    void writestring(string s) nothrow
```
Comment (mathias-lang-sociomantic) [inline src/root/outbuffer.d]: AFAICS `write` is not holding any reference to `s`, so this overload isn't needed.

Comment (mathias-lang-sociomantic) [inline src/root/outbuffer.d]: Just saw yebblies comment, nvm then.

**Comment by mathias-lang-sociomantic** [comment]
LGTM

**Comment by yebblies** [comment]
@WalterBright Please be very careful when adding member functions to D but not to the C++ headers.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Pull request merged**